src/msal/src/misc.rs

```diff
@@ -149,8 +232,141 @@ pub async fn enroll_device(graph_url: &str, access_token: &str) -> Result<Device
+    pub id: Option<String>,
+    #[serde(rename = "keyCredentials")]
+    pub key_creds: Option<Vec<KeyCredential>>,
+}
+
+#[derive(Debug, Deserialize)]
+struct ApplicationList {
+    value: Vec<Application>,
+}
+
+pub async fn list_applications(graph_url: &str, access_token: &str) -> Result<Vec<Application>> {
+    let url = &format!("{}/v1.0/applications", graph_url);
+    let client = reqwest::Client::new();
+    let resp = client
+        .get(url)
+        .header(header::AUTHORIZATION, format!("Bearer {}", access_token))
+        .send()
+        .await?;
+    if resp.status().is_success() {
+        let json_resp: ApplicationList = resp.json().await?;
+        Ok(json_resp.value)
+    } else {
+        Err(anyhow!(resp.status()))
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct KeyCredential {
+    #[serde(rename = "customKeyIdentifier")]
+    custom_key_identifier: Option<String>,
+    #[serde(rename = "displayName")]
+    display_name: Option<String>,
+    #[serde(rename = "endDateTime")]
+    end_date_time: Option<String>,
+    key: Option<String>,
+    #[serde(rename = "keyId")]
+    key_id: String,
+    #[serde(rename = "startDateTime")]
+    start_date_time: String,
+    r#type: String,
+    usage: String,
+}
+
+async fn get_application(graph_url: &str, access_token: &str, app_id: &str) -> Result<Application> {
+    let url = Url::parse_with_params(
+        &format!("{}/v1.0/applications", graph_url,),
+        &[
+            ("$select", "keyCredentials,id"),
+            ("$filter", format!("appId eq '{}'", app_id).as_str()),
+        ],
+    )?;
+    let client = reqwest::Client::new();
+    let resp = client
+        .get(url)
+        .header(header::AUTHORIZATION, format!("Bearer {}", access_token))
+        .send()
+        .await?;
+    if resp.status().is_success() {
+        let json_resp: ApplicationList = resp.json().await?;
+        match json_resp.value.first() {
+            Some(app) => Ok(app.clone()),
+            None => Err(anyhow!("Application {} not found", app_id)),
+        }
+    } else {
+        Err(anyhow!(resp.status()))
+    }
+}
+
+async fn add_application_certificate(
+    graph_url: &str,
+    access_token: &str,
+    app_id: &str,
+    cert: &str,
+    desc: &str,
+) -> Result<()> {
+    let app = get_application(graph_url, access_token, app_id).await?;
+    let app_id = match &app.id {
+        Some(id) => id.clone(),
+        None => return Err(anyhow!("Application {} missing id", app_id)),
+    };
+    let url = &format!("{}/v1.0/applications/{}", graph_url, app_id);
+    let mut key_creds = match app.key_creds {
+        Some(key_creds) => key_creds,
+        None => Vec::<KeyCredential>::new(),
+    };
+    key_creds.push(KeyCredential {
+        custom_key_identifier: None, /* Leaving this blank will default to the thumbprint */
+        display_name: Some(desc.to_string()),
+        end_date_time: None, /* Leaving this blank defaults to 1 year from now */
```

Comment:
This is the most serious concern. MS docs say that this certificate will expire after a single year. The Azure UI actually interprets this as expiring in the year 4095, so I don't know which is correct.